### PR TITLE
Add types for string-algorithms package

### PR DESCRIPTION
**Pull request by lionel-rowe**
Add types for [string-algorithms](https://github.com/burgaard/string-algorithms) package.

---

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/string-algorithms/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+// Type definitions for string-algorithms 1.0
+// Project: https://github.com/burgaard/string-algorithms
+// Definitions by: lionel-rowe <https://github.com/DefinitelyTyped>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 4.0
```
Comment (peterblazejewicz) [inline types/string-algorithms/index.d.ts:5]:
```suggestion
```
not enforced:
https://github.com/DefinitelyTyped/DefinitelyTyped#support-window

File: types/string-algorithms/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+// Type definitions for string-algorithms 1.0
+// Project: https://github.com/burgaard/string-algorithms
+// Definitions by: lionel-rowe <https://github.com/DefinitelyTyped>
```
Comment (peterblazejewicz) [inline types/string-algorithms/index.d.ts:3]: this has to be a real account handle, sorry.

File: types/string-algorithms/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        : [entries: T[], getEntry: (el: T) => number[]]
+): number[][];
+
+export function suffixArray(s: number[] | string, terminator?: number): number[];
+
+export function longestCommonPrefix(sequence: string[], suffixArray: number[]): number[];
+
+export function longestCommonSubstring(strings: string[], indexMap?: 'log' | 'linear' | StringIndexMap): string[];
+
+export class StringIndexMap {
```
Comment (peterblazejewicz) [inline types/string-algorithms/index.d.ts:21]: in terms of TS, isn't that class supposed to be `abstract`?

> Concrete implementations provide different compromises between O(1) and O(log(k)) lookup times versus O(n) and O(k) space requirements. Extend this class to implement custom mappings from string positions to substring indices with different runtime/space tradeoffs than the two pre-defined implementations.

https://github.com/burgaard/string-algorithms#class-stringindexmap

File: types/string-algorithms/index.d.ts (hunk 1 above)
Comment (lionel-rowe) [reply types/string-algorithms/index.d.ts:5]: As with the similar comment on https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63686/, removing this causes errors when I run `npm test string-algorithms` locally:

```log
Error: /root/source/repos/DefinitelyTyped/types/string-algorithms/index.d.ts:10:12
ERROR: 10:12  expect  Compile error in typescript@3.9 but not in typescript@4.0.
Fix with a comment '// Minimum TypeScript Version: 4.0' just under the header.
Cannot find name 'entries'.
[+ 7 similar errors]
/root/source/repos/DefinitelyTyped/types/string-algorithms/string-algorithms-tests.ts:7:1
ERROR: 7:1    expect  Compile error in typescript@3.9 but not in typescript@4.0.
Fix with a comment '// Minimum TypeScript Version: 4.0' just under the header.
Expected 4 arguments, but got 1.
[+ 4 similar errors]
```

In this case, they're all related to `radixSort`, which takes a variable number of parameters depending on whether the 1st parameter is already an array of `number[]` or needs to be converted first.

File: types/string-algorithms/index.d.ts (hunk 2 above)
Comment (lionel-rowe) [reply types/string-algorithms/index.d.ts]: Fixed, thanks!

File: types/string-algorithms/index.d.ts (hunk 3 above)
Comment (lionel-rowe) [reply types/string-algorithms/index.d.ts]: Yes, thanks for spotting! The class can be instantiated, but calling any of its methods just throws errors, so practically speaking it's abstract. Fixed now.

File: types/string-algorithms/index.d.ts (hunk 1 above)
Comment (peterblazejewicz) [reply types/string-algorithms/index.d.ts:5]: I see, that's named tuple. But people are already using here named tuples without TS header. Again, the same passes locally. I'd make sure latest changes from DT are in your version and local deps are reinstalled.
if issue persists, let's do not crush copies on this. Just close this comment

Comment (lionel-rowe) [reply types/string-algorithms/index.d.ts]: @peterblazejewicz re-pulling from `master` didn't fix it, but `rm -r node_modules ; npm i` did.

**Comment by peterblazejewicz** [comment]
@lionel-rowe update your branch, so the CI diff does not take the other package (which is failing now, blocking your PR)

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@lionel-rowe thanks!

**Comment by lionel-rowe** [comment]
Ready to merge

**Pull request merged**